File: Axion.Core/Compiler.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using Axion.Core.Hierarchy;
using Axion.Core.Processing.Errors;
using Axion.Core.Processing.Lexical;
using Axion.Core.Processing.Syntactic;
using Axion.Core.Processing.Syntactic.Expressions;
using Axion.Core.Processing.Translation;
using Axion.Core.Processing.Traversal;
using Axion.Core.Specification;
using NLog;
using Module = Axion.Core.Hierarchy.Module;

namespace Axion.Core {
    public class Compiler {
        private static readonly Assembly coreAsm =
            Assembly.GetExecutingAssembly();

        public static readonly string Version =
            coreAsm.GetName().Version.ToString();

        /// <summary>
        ///     Path to directory where compiler executable is located.
        /// </summary>
        public static readonly string WorkDir =
            AppDomain.CurrentDomain.BaseDirectory;

        /// <summary>
        ///     Path to directory where generated output is located.
        /// </summary>
        public static readonly string OutDir = Path.Join(WorkDir, "output");

        public static readonly Dictionary<string, INodeTranslator> Translators = new();

        private static readonly Logger logger =
            LogManager.GetCurrentClassLogger();

        public static string GetTempSourceFilePath() {
            return Path.Combine(
                WorkDir,
                "temp",
                "tmp-" + DateTime.Now.ToFileName() + Spec.FileExtension
            );
        }

        public static void AddTranslator(
            string          name,
            INodeTranslator translator
        ) {
            Translators[name.Trim().ToLower()] = translator;
        }

        public static object? Process(
            Project           project,
            ProcessingOptions options
        ) {
            logger.Debug($"Processing project '{project.ConfigFile.Name}'");
            var result = Process(project.MainModule, options);
            return result;
        }

        public static object? Process(
            Module            module,
            ProcessingOptions options
        ) {
            logger.Debug(
                module.Parent == null
                    ? $"Processing module '{module.Name}'"
                    : $"Processing submodule '{module.Name}'"
            );
            object? result = null;
            foreach (var (_, subModule) in module.Submodules) {
                result = Process(subModule, options);
            }

            foreach (var (_, unit) in module.Units) {
                result = Process(unit, options);
            }

            return result;
        }

        public static object? Process(Unit src, ProcessingOptions options) {
            logger.Debug($"Processing '{src.SourceFile.Name}'");
            if (src.TextStream.IsEmpty) {
                logger.Error("Source is empty. Processing aborted.");
                return null;
            }

            object? result = null;
            foreach (var step in CompilationSteps) {
                result = step.Value(src, options);
                if (options.ProcessingMode == step.Key || src.HasErrors) {
                    break;
                }
            }

            var errCount = 0;
            foreach (var e in src.Blames) {
                logger.Error(e.ToString());
                if (e.Severity == BlameSeverity.Error) {
                    errCount++;
                }
            }

            logger.Debug(
                errCount > 0 ? "Processing aborted." : "Processing completed."
            );

            return result;
        }

        public static readonly Dictionary<Mode, Func<Unit, ProcessingOptions, object>>
            CompilationSteps = new() {
                { Mode.Lexing, Lex },
                { Mode.Parsing, Parse },
                { Mode.Reduction, Reduce },
                { Mode.Translation, Translate }
            };

        public static TokenStream Lex(Unit src, ProcessingOptions options) {
            logger.Debug("Tokens list generation");
            var lexer = new Lexer(src);
            while (true) {
                var token = lexer.Read();
                if (token == null) {
                    continue;
                }

                src.TokenStream.Add(token);
                if (lexer.ProcessTerminators.Contains(token.Type)) {
                    break;
                }
            }

            foreach (var mismatch in lexer.MismatchingPairs) {
                LanguageReport.MismatchedBracket(mismatch);
            }

            return src.TokenStream;
        }

        private static Ast Parse(Unit src, ProcessingOptions options) {
            logger.Debug("Abstract Syntax Tree generation");
            src.Ast.Parse();
            return src.Ast;
        }

        private static Ast Reduce(Unit src, ProcessingOptions options) {
            logger.Debug("Syntax tree reducing");
            Traversing.Traverse(src.Ast);
            return src.Ast;
        }

        private static CodeWriter Translate(
            Unit              src,
            ProcessingOptions options
        ) {
            if (!Translators.TryGetValue(
                options.TargetLanguage,
                out var ncv
            )) {
                logger.Error($"No frontend with name {options.TargetLanguage}");
                return CodeWriter.Default;
            }

            var cw = new CodeWriter(ncv);
            try {
                cw.Write(src.Ast);
                var code = cw.ToString();
                logger.Debug(code);
                File.WriteAllText(
                    Path.Combine(
                        src.OutputDirectory.FullName,
                        Path.ChangeExtension(
                            src.SourceFile.Name,
                            cw.OutputFileExtension
                        )
                    ),
                    code
                );
            }
            catch (Exception ex) {
                logger.Error("Translation failed:");
                logger.Info(ex.Message);
            }

            return cw;
        }
    }
}

File: compiler.py
import logging
from datetime import datetime
from os.path import dirname, realpath
from pathlib import Path
from pprint import pprint

from anytree import RenderTree
from anytree.render import ContStyle

from axion import logger
from processing.codegen.code_builder import OutputLang, CodeBuilder
from processing.mode import ProcessingMode
from processing.options import ProcessingOptions
from source_unit import SourceUnit
from utils import resolve_path


class Compiler:
    version = "0.4.0.0-alpha"

    source_file_ext = ".ax"
    output_file_ext = ".out"
    debug_file_ext = ".dbg.json"

    compiler_dir = Path(dirname(realpath(__file__)))

    @staticmethod
    def temp_source_path() -> Path:
        path = Path(
            Compiler.compiler_dir / "temp" /
            ("tmp-" + datetime.now().strftime("%b-%d_%H-%M-%S-%f") + Compiler.source_file_ext).lower()
        )
        return resolve_path(path)

    @staticmethod
    def process_source(
            source: SourceUnit,
            mode: ProcessingMode = ProcessingMode.compile,
            options: ProcessingOptions = ProcessingOptions.default
    ):
        logger.info(f"Processing '{source.source_path.name}'")

        def process():
            Compiler.lexical_analysis(source)
            if logger.isEnabledFor(logging.DEBUG) and False:
                Compiler.lexical_debug_output(source)

            if mode == ProcessingMode.lex:
                return

            Compiler.syntax_parsing(source)
            if logger.isEnabledFor(logging.DEBUG):
                Compiler.syntax_debug_output(source)

            if mode == ProcessingMode.parsing:
                return

            transpiled = Compiler.transpile(source, OutputLang.csharp)
            logger.debug(f"-- Generated code")
            print(transpiled)

        process()
        logger.info('-- Errors')
        pprint(source.blames)

    @staticmethod
    def lexical_analysis(source: SourceUnit):
        from processing.lexical.tokens.token import Token
        from errors.blame import BlameType

        logger.debug('-- Lexical analysis')
        while True:
            token = Token(source = source).read()
            if token is not None:
                source.token_stream.tokens.append(token)
                if token.ttype in source.process_terminators:
                    break
        for mismatch in source.mismatching_pairs:
            source.blame(
                BlameType[
                    'mismatched_' + str(mismatch.ttype)
                        .replace('TokenType.', '')
                        .replace('open_', '')
                        .replace('close_', '')
                    ],
                mismatch
            )

    @staticmethod
    def lexical_debug_output(source: SourceUnit):
        logger.debug('-- Tokens')
        print('\n'.join([repr(t) for t in source.token_stream.tokens]))

    @staticmethod
    def syntax_parsing(source: SourceUnit):
        logger.debug('-- Syntax parsing')
        source.ast.parse()

    @staticmethod
    def syntax_debug_output(source: SourceUnit):
        logger.debug('-- AST')
        print(RenderTree(source.ast, ContStyle()))

    @staticmethod
    def transpile(source: SourceUnit, out_lang: OutputLang) -> str:
        logger.debug(f"-- Code generation for '{out_lang.name}' target")
        builder = CodeBuilder(out_lang)
        builder += source.ast
        default_imports = [
            'System',
            'System.IO',
            'System.Linq',
            'System.Text',
            'System.Numerics',
            'System.Threading',
            'System.Diagnostics',
            'System.Collections',
            'System.Collections.Generic'
        ]
        using_str = ''
        for using in default_imports:
            using_str += f'using {using};\n'
        return using_str + '\n' + builder.code
